File: e2e/clients/occurrence.py
from typing import Dict, Any

import api
from e2e.common import base_url, build_query_string
from shared import constants
occurrence_path_get = base_url + '/occurrence',
occurrence_path_update = base_url + '/occurrence/{id}',
occurrence_path_create = base_url + '/task/{task_id}/occurrence',


def create(task_id: int, priority: int, jwt: str) -> int:
    return api.create(
        occurrence_path_create.format(task_id=task_id), {
            constants.priority: priority,

        }, jwt)


def edit(id: int, priority: int, jwt: str):
    return api.edit(
        occurrence_path_update.format( id=id), {
            constants.priority: priority,
        }, jwt)


def delete(id: int, jwt: str):
    return api.delete(occurrence_path_update.format(id=id), jwt)


def get(id: int, jwt: str, query_params: Dict[str, str] = {}, ) -> Dict[str, Any]:
    return api.get((occurrence_path_update.format(id=id) if id else occurrence_path_get) + f'?{build_query_string(query_params)}', jwt)

File: mobile/lib/services/occurrence.dart
import '../common/constants.dart';
import '../common/models.dart';
import 'api_client.dart';

const String _path = '/occurrence';

class OccurrenceService {
  final ApiClient _apiClient = ApiClient();

  Future<Map<String, dynamic>> create(
    int taskId,
    int priority, {
    int? time,
    bool? completed,
  }) async {
    return await _apiClient.post(
      '/task/$taskId/occurrence',
      body: {
        kPriority: priority,
        if (time != null) kTime: time,
        if (completed != null) kCompleted: completed,
      },
    );
  }

  Future<List<Occurrence>> list({Map<String, String>? queryParams}) async {
    final response = await _apiClient.get(_path, queryParams: queryParams);
    return response
        .map((occurrence) => Occurrence.fromJson(occurrence))
        .toList();
  }

  Future<Occurrence> get(int id) async {
    final response = await _apiClient.get('$_path/$id');
    return Occurrence.fromJson(response.first);
  }

  Future<void> update(
    int id, {
    int? priority,
    int? time,
    bool? completed,
  }) async {
    final body = {
      if (priority != null) kPriority: priority,
      if (completed != null) kCompleted: completed,
      if (time != null) kTime: time,
    };
    await _apiClient.patch('$_path/$id', body: body);
  }

  Future<void> delete(int id) async {
    await _apiClient.delete('$_path/$id');
  }
}
